new-log-viewer/src/components/Editor/MonacoInstance/index.tsx

```diff
@@ -0,0 +1,127 @@
+import {
+    useEffect,
+    useRef,
+} from "react";
+
+import * as monaco from "monaco-editor/esm/vs/editor/editor.api";
+
+import {ActionType} from "../../../utils/actions";
+import {
+    CursorExplicitPosChangeCallback,
+    CustomActionCallback,
+} from "./typings";
+import {
+    goToPositionAndCenter,
+    initMonacoEditor,
+} from "./utils";
+
+import "./index.css";
+
+
+interface MonacoEditorProps {
+    lineNum: number,
+    text: string,
+    actions: ActionType[],
+    beforeMount?: () => void,
+    beforeTextUpdate?: (editor: monaco.editor.IStandaloneCodeEditor) => void,
+    onCursorExplicitPosChange: CursorExplicitPosChangeCallback,
+    onCustomAction: CustomActionCallback,
+    onMount?: (editor: monaco.editor.IStandaloneCodeEditor) => void,
+    onTextUpdate?: (editor: monaco.editor.IStandaloneCodeEditor) => void,
+}
+
+/**
+ * Wraps a `monaco-editor` instance created from the DOM rendered, which accepts a variety of props
+ * to configure text content, custom actions, and various lifecycle hooks for interacting with
+ * the editor.
+ *
+ * @param props
+ * @param props.text
+ * @param props.lineNum
+ * @param props.actions
+ * @param props.beforeMount
+ * @param props.beforeTextUpdate
+ * @param props.onCursorExplicitPosChange
+ * @param props.onCustomAction
+ * @param props.onMount
+ * @param props.onTextUpdate
+ * @return
+ */
+const MonacoEditor = ({
+    lineNum,
+    text,
+    actions,
+    beforeMount,
+    beforeTextUpdate,
+    onMount,
+    onCursorExplicitPosChange,
+    onCustomAction,
+    onTextUpdate,
+}: MonacoEditorProps) => {
+    const editorRef = useRef<null|monaco.editor.IStandaloneCodeEditor>(null);
+    const editorContainerRef = useRef<HTMLDivElement>(null);
+    const lineNumRef = useRef<number>(lineNum);
+
+    // Synchronize `lineNumRef` with `lineNum`.
+    useEffect(() => {
+        lineNumRef.current = lineNum;
+    }, [lineNum]);
+
+    useEffect(() => {
+        console.log("Initiating monaco instance");
+        if (null === editorContainerRef.current) {
+            console.error("Unexpected unmounted editor container div element");
```

Comment:
Won't this log be printed the first time the component mounts?

Reply:
Supposedly never. The editorContainerRef is for referencing the `div` DOM node "monaco-container". When a `useEffect(()=>{...}, [])` hook gets invoked in an FC / a `componentDidMount` method gets invoked in a Class Component, the DOM nodes should have been mounted so we do not anticipate the `ref.current` to a hold a null value.

So the log here is for docs purpose (and just in case something unexpected really happens).

Reply:
True, true.